guppylang/checker/stmt_checker.py

```diff
@@ -46,17 +46,50 @@ def _check_expr(
     ) -> tuple[ast.expr, Subst]:
         return ExprChecker(self.ctx).check(node, ty, kind)
 
-    def _check_assign(self, lhs: ast.expr, ty: Type, node: ast.stmt) -> None:
+    def _check_assign(self, lhs: ast.expr, ty: Type, node: ast.stmt) -> ast.expr:
         """Helper function to check assignments with patterns."""
         match lhs:
             # Easiest case is if the LHS pattern is a single variable.
             case ast.Name(id=x):
-                # Store the type in the AST
-                with_type(ty, lhs)
-                self.ctx.locals[x] = Variable(x, ty, lhs)
+                var = Variable(x, ty, lhs)
+                self.ctx.locals[x] = var
+                return with_loc(lhs, PlaceNode(place=var))
+
+            case ast.Attribute(value=value, attr=attr):
+                value, struct_ty = self._synth_expr(value)
+                if (
+                    not isinstance(struct_ty, StructType)
+                    or attr not in struct_ty.field_dict
+                ):
+                    raise GuppyTypeError(
+                        f"Expression of type `{struct_ty}` has no attribute `{attr}`",
+                        # Unfortunately, `attr` doesn't contain source annotations, so
+                        # we have to use `lhs` as the error location
+                        lhs,
+                    )
+                field = struct_ty.field_dict[attr]
+                # TODO: In the future, we could infer some type args here
+                if field.ty != ty:
+                    raise GuppyTypeError(
+                        f"Cannot assign expression of type `{ty}` to field with type "
+                        f"`{field.ty}`",
+                        lhs,
+                    )
+                if not isinstance(value, PlaceNode):
+                    # For now we complain if someone tries to assign to something that
+                    # is not a place (e.g. `f().a = 4`)
+                    raise GuppyError(
+                        "Assigning to this expression is not supported yet", value
```

Comment:
worth more details on why not, otherwise I don't necessarily know how to fix it

Reply:
I meant more details in the error message

Reply:
Ops yeah I misunderstood. I added a fix suggestion to the error message